packages/workflow/src/workflow.ts

```diff
@@ -1356,21 +1384,119 @@ export function upsertSearchAttributes(searchAttributes: SearchAttributes): void
     throw new Error('searchAttributes must be a non-null SearchAttributes');
   }
 
-  activator.pushCommand({
-    upsertWorkflowSearchAttributes: {
-      searchAttributes: mapToPayloads(searchAttributePayloadConverter, searchAttributes),
-    },
-  });
+  if (Array.isArray(searchAttributes)) {
+    // Typed search attributes
+    activator.pushCommand({
+      upsertWorkflowSearchAttributes: {
+        searchAttributes: mapToPayloads<string, TypedSearchAttributeUpdateValue<SearchAttributeType>>(
+          typedSearchAttributePayloadConverter,
+          Object.fromEntries(
+            searchAttributes.map((pair) => [
+              pair.key.name,
+              new TypedSearchAttributeUpdateValue(pair.key.type, pair.value),
+            ])
+          )
+        ),
+      },
+    });
 
-  activator.mutateWorkflowInfo((info: WorkflowInfo): WorkflowInfo => {
-    return {
-      ...info,
-      searchAttributes: {
-        ...info.searchAttributes,
-        ...searchAttributes,
+    activator.mutateWorkflowInfo((info: WorkflowInfo): WorkflowInfo => {
+      // Create a copy of the current state.
+      const newSearchAttributes: SearchAttributes = { ...info.searchAttributes };
+      for (const pair of searchAttributes) {
+        if (pair.value === null) {
```

Comment:
Here, we check if value is `null`, but in legacy SA, we check for it to be `undefined`. That may be confusing. Let's use `== null` in both cases, just in case.

Reply:
Changed to `== null`.